### DISCUSSION TITLE
handlers and combined effects

### DISCUSSION THREAD

**Discussion by dcurrie** [koka-lang/koka#133] [Q&A] [answered]
It surprised me that nesting two effect handlers let one of the effects leak... (I have no certainty that my terminology here or interpretation of what is really happening is correct, hence in Q&A section!)... 

My goal is to write a unit test that has a mutable vector and uses random numbers. Each of those two things work separately for me. When I combine them, I get
```
(1, 0): error: there are unhandled effects for the main expression
  inferred effect : <pure,console,rstest/istate<int>>
  unhandled effect: rstest/istate<int>
  hint            : wrap the main function in a handler
```
and I have no idea how to wrap the main function in a handler. I found a workaround, but I don't understand it. 

This fails:
```koka
fun istate-test(): <exn,div,random,istate<int>> string {
    istate(100, -9, show-a-couple-prng)
}
public fun main() {
  println("Random numbers " + pseudo-random(7, istate-test))
}
```
This fails:
```koka
public fun main() {
  fun istate-test(): <exn,div,random,istate<int>> string {
    istate(100, -9, show-a-couple-prng)
  }
  println("Random numbers " + pseudo-random(7, istate-test))
}
```
This works!?:
```koka
public fun main() {
  println("Random numbers " + pseudo-random(7, fn() { istate(100, -9, show-a-couple-prng) }))
}
```
Full test code is [here](https://github.com/dcurrie/koka-bbtree/blob/main/rstest.kk)

**Comment by anfelor** [comment, marked as answer]
I think your type signature for `istate-test` is wrong: 
```fun istate-test(): <exn,div,random,istate<int>> string```
still has the `istate` effect. I think if you remove that effect from the type signature you should be fine.
In particular, if you add an effect to the type signature that is not needed in the function, Koka will subsequently assume that the function needed the effect.

**Comment by dcurrie** [comment]
Thanks, Anton! That fixes it. I was confused by the error message.
